### PR TITLE
Remove EntitySource

### LINKED ISSUE #97: Remove the EntitySource API and replace it with the VariableSource API
At a high level, the Deppy Library checks if a set of constraints can be satisfied by a set of entities. Today, the Deppy Library expects users to implement the [EntitySource interface](https://pkg.go.dev/github.com/operator-framework/deppy/pkg/deppy/input#EntitySource) to provide the entities that Deppy considers. There are a number of shortcomings with the existing interface that justify moving to using the VariableSource interface.
1.  **This API doesn't provide value:** Deppy should not care if a user has given it an EntitySource, it should only rely on a set of Variables being provided, which are retrievable through the existing VariableSource interface
2. **Needless conversion:**: When Deppy is given a EntitySource, it is converted into Entities, [which are further converted into variables for resolution](https://github.com/operator-framework/deppy/blob/9a7655cd364b3900d394f81a752731200e2d01a4/pkg/deppy/solver/solver.go#L94). After resolution, we again [convert the variables to entities](https://github.com/operator-framework/operator-controller/blob/5cc5cc26692b3d71fa870d62c3344dd44300fd08/controllers/operator_controller.go#L133)  to get the final list of output. The flow looks like:   `Input -> Entities -> Variables -> Solver -> Solution set (variables) -> Entities`.

Moving away from EntitySources / entities will yield the following benefits:
- Simplification of the codebase.
- The EntitySource interface is strongly tied to the bundle concept introduced by OLM. Moving away from EntitySources/Entites would allow us to support other formats.

Open Questions:
- [spike] Do we need separate variable sources for each constraint? ([eg](https://pkg.go.dev/github.com/operator-framework/operator-controller/internal/resolution/variable_sources/crd_constraints#NewCRDUniquenessConstraintsVariableSource)). How would OLM specific constraints look like?

Follow up:
- Currently we have [BundleEntities](https://pkg.go.dev/github.com/operator-framework/operator-controller/internal/resolution/variable_sources/entity#BundleEntity) in operator controller, these need to be converted in variable sources and directly fed to resolver. The GetVariables method would take in the BMD and packages, construct a valid deppy identifier and add it to the resolver [here](https://github.com/operator-framework/deppy/pull/95/files#diff-446aa23ef0c712f68e80aa91316ed79ac642b4765af71c489a839ad953131eabR27-R29)
- Catalogd should also be updated to use VariableSources.

Related PRs: 
- https://github.com/operator-framework/deppy/pull/95

Demo Script:
- Is there a need for a demo script?

### PR DESCRIPTION
**Pull request by perdasilva**
This PR replaces #95 by rebasing it and fixing merge conflicts. The intent of this PR is to remove the concepts of EntitySource and Entity from Deppy. The `VariableSource` interface can be implemented with access to specific backend clients (e.g. registry, kube client, etc.) in order to provide their `Variable`s.

Closes #97

### CODE REVIEW

**Comment by ncdc** [comment]
Is there some background on this change?

**Comment by ncdc** [comment]
I take it this is for #97? And it presumably replaces #95?

**Comment by perdasilva** [comment]
Sorry @ncdc - I started this just before I left for holidays. Yeah, the idea was to take Varsha's PR, rebase, fix and put it back here. I'll fill out the description and tie it to a ticket.

**Review by m1kola — COMMENTED** [review]
The change looks good to me apart from removal of `rand.Seed`.

Updating operator-controller will be fun.  And I'll be missing predicates.

File: cmd/sudoku/sudoku.go [hunk 1]
```diff
@@ -27,29 +22,13 @@ func GetID(row int, col int, num int) deppy.Identifier {
...
-	}
+	return &Sudoku{}
 }
 
-func (s Sudoku) GetVariables(_ context.Context, _ input.EntitySource) ([]deppy.Variable, error) {
+func (s Sudoku) GetVariables(_ context.Context) ([]deppy.Variable, error) {
 	// adapted from: https://github.com/go-air/gini/blob/871d828a26852598db2b88f436549634ba9533ff/sudoku_test.go#L10
 	variables := make(map[deppy.Identifier]*input.SimpleVariable, 0)
 	inorder := make([]deppy.Variable, 0)
-	rand.Seed(time.Now().UnixNano())
```
[the remark this answers is not in the document]
Comment (joelanford) [reply cmd/sudoku/sudoku.go:52]: I just merged the Go 1.20 bump. Need to rebase here now, it looks like.

**Comment by tmshort** [comment]
Requires another rebase...

**Review by m1kola — APPROVED** [review]
Looks good to me!

**Pull request merged**